File: 2016/python/03.py
# Advent of Code 2016, day 3
# (c) aichingert

def is_triangle(a,b,c):
    return a+b>c and a+c>b and b+c>a

def part1(inp):
    ans = 0
    for l in inp:
        vls = list(map(int, list(filter(lambda x: x != '', l.split(' ')))))
        if is_triangle(vls[0], vls[1], vls[2]):
            ans += 1
    return ans

def part2(inp):
    ans = 0
    for i in range(0,len(inp), 3):
        f = list(map(int, list(filter(lambda x: x != '', inp[i].split(' ')))))
        s = list(map(int, list(filter(lambda x: x != '', inp[i+1].split(' ')))))
        t = list(map(int, list(filter(lambda x: x != '', inp[i+2].split(' ')))))
        if is_triangle(f[0],s[0],t[0]): ans += 1
        if is_triangle(f[1],s[1],t[1]): ans += 1
        if is_triangle(f[2],s[2],t[2]): ans += 1
    return ans

inp = open("../input/03").read().split('\n')
print("Part 1:",part1(inp))
print("Part 2:",part2(inp))

File: old/2016/rust/03.rs
// Advent of Code 2016, day 3
// (c) aichingert

fn valid_triangle(ver: &Vec<i32>) -> bool {
    ver[0] + ver[1] > ver[2] && ver[1] + ver[2] > ver[0] && ver[0] + ver[2] > ver[1]
}

fn part1(inp: &Vec<Vec<i32>>) -> usize {
    inp.iter().filter(|v| valid_triangle(v)).count()
}

fn part2(inp: &Vec<Vec<i32>>) -> usize {
    let mut ans: usize = 0;

    for i in (0..inp.len()).step_by(3) {
        ans += [[inp[i][0],inp[i+1][0],inp[i+2][0]],[inp[i][1],inp[i+1][1],inp[i+2][1]],[inp[i][2],inp[i+1][2],inp[i+2][2]]]
            .iter()
            .filter(|v| valid_triangle(&v.to_vec()))
            .count();
    }

    ans
}

fn main() {
    let inp = std::fs::read_to_string("../input/03").unwrap();
    let inp = inp.lines().map(|l| l.split(' ').filter(|s| !s.is_empty()).map(|s| s.parse::<i32>().unwrap()).collect::<Vec<i32>>()).collect::<Vec<Vec<i32>>>();

    println!("Part 1: {}", part1(&inp));
    println!("Part 2: {}", part2(&inp));
}
